_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+```
+
+And add Spring Cloud Gateway with Resilience4j dependencies:
+
+```groovy
+implementation 'org.springframework.cloud:spring-cloud-starter-circuitbreaker-reactor-resilience4j'
+implementation 'org.springframework.cloud:spring-cloud-starter-gateway'
+```
+
+Then, I moved `CoolCarController` to its own class and re-implemented it with WebClient.
+
+```java
+package com.example.apigateway.web;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.client.circuitbreaker.ReactiveCircuitBreaker;
+import org.springframework.cloud.client.circuitbreaker.ReactiveCircuitBreakerFactory;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Flux;
+
+@RestController
+class CoolCarController {
+
+    Logger log = LoggerFactory.getLogger(CoolCarController.class);
+
+    private final WebClient.Builder webClientBuilder;
+    private final ReactiveCircuitBreaker circuitBreaker;
+
+    public CoolCarController(WebClient.Builder webClientBuilder,
+                             ReactiveCircuitBreakerFactory circuitBreakerFactory) {
+        this.webClientBuilder = webClientBuilder;
+        this.circuitBreaker = circuitBreakerFactory.create("circuit-breaker");
+    }
+
+    record Car(String name) {
+    }
+
+    @GetMapping("/cool-cars")
+    public Flux<Car> coolCars() {
+        return circuitBreaker.run(
+            webClientBuilder.build()
+                .get().uri("http://car-service/cars")
+                .retrieve().bodyToFlux(Car.class)
+                .filter(this::isCool),
+            throwable -> {
+                log.warn("Error making request to car service", throwable);
+                return Flux.empty();
+            });
+    }
+
+    private boolean isCool(Car car) {
+        return !car.name().equals("AMC Gremlin") &&
+            !car.name().equals("Triumph Stag") &&
+            !car.name().equals("Ford Pinto") &&
+            !car.name().equals("Yugo GV");
+    }
+}
+```
+
+In the `car-service` project, I had to switch from using Spring Data REST to handling it with a `@RestController` and `@GetMapping` annotation. I removed the `@RepositoryRestResource` annotation from `CarRepository` and added a `CarController` class.
+
+```java
+package com.example.carservice.web;
+
+import com.example.carservice.data.Car;
+import com.example.carservice.data.CarRepository;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+class CarController {
+
+    private final CarRepository repository;
+
+    public CarController(CarRepository repository) {
+        this.repository = repository;
+    }
+
+    @GetMapping("/cars")
+    public List<Car> getCars() {
+        return repository.findAll();
+    }
+}
+```
+
```

Comment:
**NOTE**: I did try to use Spring HATEOAS but [ran into an issue](https://github.com/okta/okta-spring-boot/issues/607) when using it with the Okta Spring Boot starter.